django/db/models/expressions.py

```diff
@@ -274,23 +284,26 @@ def _resolve_output_field(self):
                 raise FieldError('Expression contains mixed types. You must set output_field.')
             return output_field
 
-    def convert_value(self, value, expression, connection):
+    @staticmethod
+    def _convert_value_noop(value, expression, connection):
+        return value
+
+    @cached_property
+    def convert_value(self):
         """
         Expressions provide their own converters because users have the option
         of manually specifying the output_field which may be a different type
         from the one the database returns.
         """
         field = self.output_field
         internal_type = field.get_internal_type()
-        if value is None:
-            return value
-        elif internal_type == 'FloatField':
-            return float(value)
+        if internal_type == 'FloatField':
+            return lambda value, expression, connection: None if value is None else float(value)
```

Comment:
@sir-sigurd these lambdas break expression pickling:
```
Traceback (most recent call last):
  File "/home/travis/build/millerdev/django-cte/tests/test_recursive.py", line 212, in test_pickle_recursive_cte_queryset
    pickled_qs = pickle.loads(pickle.dumps(regions))
AttributeError: Can't pickle local object 'BaseExpression.convert_value.<locals>.<lambda>'
```
Found here: https://travis-ci.org/millerdev/django-cte/jobs/366128465

Reply:
Please report bugs at https://code.djangoproject.com/.

Reply:
False alarm. Looks like the [custom `__getstate__`](https://github.com/django/django/pull/8854/files#diff-fe637a3ec6af9cbe2ee87293000e7c82R150) has subsequently been removed, which is the likely cause of the issue. Tracking that down now.